hknweb/candidate/views.py

```diff
@@ -467,6 +467,109 @@ def add_cands(request):
+        raise Http404()
+    next_page = request.POST.get("next", '/')
+    csv_file = request.FILES.get("csv_file", None)
+    if not csv_file or not csv_file.name.endswith(ATTR.CSV_ENDING):
+        messages.error(request, "Please input a csv file!")
+        return redirect(next_page)
+    decoded_csv_file = csv_file.read().decode(ATTR.UTF8).splitlines()
+    mem_csv = csv.DictReader(decoded_csv_file)
+
+    checkoff_type = request.POST.get("checkoff_type", "")
+    if checkoff_type == "event":
+        event_id = request.POST.get("event_id", "")
+        if not event_id:
+            messages.error(request, "Please input an event ID!")
+            return redirect(next_page)
+        event = Event.objects.filter(pk=event_id)
+        if not event:
+            messages.error(request, "Please input a valid event ID!")
+            return redirect(next_page)
+        event = event[0]
+    elif checkoff_type == "project":
+        project_name = request.POST.get("project_name", "")
+        project = CommitteeProject.objects.get(name=project_name)
+        projectDoneEntry = CommitteeProjectDoneEntry.objects.filter(committeeProject=project).first()
+        if projectDoneEntry is None:
+            messages.error(request, "Could not find a corresponding CommiteeProjectDoneEntry. Please make sure one is created for the project.")
+            return redirect(next_page)
+    elif checkoff_type == "dues":
+        dues_name = request.POST.get("dues_name", "")
+        due = DuePayment.objects.get(name=dues_name)
+        duesDoneEntry = DuePaymentPaidEntry.objects.filter(duePayment=due).first()
+        if duesDoneEntry is None:
+            messages.error(request, "Could not find a corresponding DuePaymentPaidEntry. Please make sure one is created for the due.")
+            return redirect(next_page)
+    elif checkoff_type == "forms":
+        forms_name = request.POST.get("forms_name", "")
+        form = CandidateForm.objects.get(name=forms_name)
+        formsDoneEntry = CandidateFormDoneEntry.objects.filter(form=form).first()
+        if formsDoneEntry is None:
+            messages.error(request, "Could not find a corresponding CandidateFormDoneEntry. Please make sure one is created for the form.")
+            return redirect(next_page)
+    
+
+    # Pre-screen and validate data
+    users = []
+    for row in mem_csv:
+        try:
+            memberdto = CandidateDTO(row)
+        except AssertionError as e:
+            messages.error(request, "Invalid CSV format. Check that your columns are correctly labeled and filled out for each row.")
+            return redirect(next_page)
+        user = User.objects.filter(first_name=memberdto.first_name, last_name=memberdto.last_name, email=memberdto.email)
+        if not user:
+            messages.error(request, "Could not find user " + memberdto.first_name + " " + memberdto.last_name + " with email " + memberdto.email + ". Please check these parameters again.")
```

Comment:
Let's add a part here that says: No checkoff actions have been taken. Fix the errors and re-upload the entire file.